go/chat/storage/inbox.go

```diff
@@ -699,76 +722,75 @@ func (i *Inbox) NewConversation(ctx context.Context, uid gregor1.UID, vers chat1
-				if iconv.GetConvID().Eq(super.ConversationID) {
-					i.Debug(ctx, "NewConversation: setting supersededBy: target: %s superseder: %s",
-						iconv.ConvIDStr, conv.GetConvID())
-					iconv.Conv.Metadata.SupersededBy = append(iconv.Conv.Metadata.SupersededBy, conv.Metadata)
-					iconv.Conv.Metadata.Version = vers.ToConvVers()
+		if len(conv.Metadata.Supersedes) > 0 {
+			for _, convID := range iboxIndex.ConversationIDs {
+				iconv, err := i.readConv(ctx, uid, convID)
+				if err != nil {
+					return err
+				}
+				if iconv.Conv.Metadata.FinalizeInfo == nil {
+					continue
+				}
+				for _, super := range conv.Metadata.Supersedes {
+					if iconv.GetConvID().Eq(super.ConversationID) {
+						i.Debug(ctx, "NewConversation: setting supersededBy: target: %s superseder: %s",
+							iconv.ConvIDStr, conv.GetConvID())
+						iconv.Conv.Metadata.SupersededBy = append(iconv.Conv.Metadata.SupersededBy, conv.Metadata)
+						iconv.Conv.Metadata.Version = vers.ToConvVers()
+						if err := i.writeConv(ctx, uid, iconv); err != nil {
+							return err
+						}
+					}
 				}
 			}
 		}
-
+		if err := i.writeConv(ctx, uid, utils.RemoteConv(conv)); err != nil {
+			return err
+		}
 		// only chat convs for layout changed
 		layoutChanged = conv.GetTopicType() == chat1.TopicType_CHAT
-		ibox.Conversations = append(utils.RemoteConvs([]chat1.Conversation{conv}), ibox.Conversations...)
+		iboxIndex.ConversationIDs = append(iboxIndex.ConversationIDs, conv.GetConvID())
+		if err := i.writeDiskIndex(ctx, uid, iboxIndex); err != nil {
+			return err
+		}
 	} else {
 		i.Debug(ctx, "NewConversation: skipping update, conversation exists in inbox")
 	}
 
 	// Write out to disk
-	ibox.InboxVersion = vers
-	return i.writeDiskInbox(ctx, uid, ibox)
-}
-
-func (i *Inbox) getConv(convID chat1.ConversationID, convs []types.RemoteConversation) (int, *types.RemoteConversation) {
-
-	var index int
-	var conv types.RemoteConversation
-	found := false
-	for index, conv = range convs {
-		if conv.GetConvID().Eq(convID) {
-			found = true
-			break
-		}
-	}
-	if !found {
-		return 0, nil
-	}
-
-	return index, &convs[index]
+	iboxVers.InboxVersion = vers
+	return i.writeDiskVersions(ctx, uid, iboxVers)
 }
 
 // Return pointers into `convs` for the convs belonging to `teamID`.
-func (i *Inbox) getConvsForTeam(ctx context.Context, teamID keybase1.TeamID, convs []types.RemoteConversation) (res []*types.RemoteConversation) {
+func (i *Inbox) getConvsForTeam(ctx context.Context, uid gregor1.UID, teamID keybase1.TeamID,
+	index inboxDiskIndex) (res []types.RemoteConversation) {
 	tlfID, err := chat1.TeamIDToTLFID(teamID)
 	if err != nil {
 		i.Debug(ctx, "getConvsForTeam: teamIDToTLFID failed: %v", err)
 		return nil
 	}
-	for i := range convs {
-		conv := &convs[i]
-		if conv.Conv.GetMembersType() == chat1.ConversationMembersType_TEAM && conv.Conv.Metadata.IdTriple.Tlfid.Eq(tlfID) {
+	for _, convID := range index.ConversationIDs {
+		conv, err := i.readConv(ctx, uid, convID)
```

Comment:
bummer we have to read everything here, how often is this method called? maybe we could keep an index for convs by tlf id too?

Reply:
Very rarely, only if the team wide retention is set.